### PR TITLE
Stabilize inline asm for LoongArch64

### PR DESCRIPTION
**Pull request by heiher**
This PR is used to tracking for stabilize `inline asm` for LoongArch64.

**Status**

- [x] https://github.com/rust-lang/rust/pull/111237
- [x] https://github.com/rust-lang/rust/pull/111332
- [x] https://github.com/rust-lang/reference/pull/1357

Any others I missed?

r? @Amanieu

### CODE REVIEW

File: compiler/rustc_target/src/asm/mod.rs [hunk 1]
```diff
@@ -1022,6 +1027,21 @@ impl InlineAsmClobberAbi {
...
+                    r4, r5, r6, r7, r8, r9, r10, r11,
+                    // t0-t8
+                    r12, r13, r14, r15, r16, r17, r18, r19, r20,
+                    // fa0-fa7
+                    f0, f1, f2, f3, f4, f5, f6, f7,
+                    // ft0-ft15
+                    f8, f9, f10, f11, f12, f13, f14, f15,
+                    f16, f17, f18, f19, f20, f21, f22, f23,
+                }
+            },
```
Comment (workingjubilee) [inline compiler/rustc_target/src/asm/mod.rs:1044]: If you have improvements to inline `asm!` for LoongArch, please land them in a separate PR.

**Review by workingjubilee — COMMENTED** [review]
A stabilization PR should have no _functional_ changes, because it's already under review for correctness in changing the stability gates and feature flags appropriately.

File: compiler/rustc_target/src/asm/mod.rs (hunk 1 above)
Comment (heiher) [reply compiler/rustc_target/src/asm/mod.rs:1044]: I see. Thank you.

**Comment by Amanieu** [comment]
I just did another full review of the LoongArch64 inline asm implementation. The main thing that stood out was that the behavior of `preserves_flags` is not defined. LoongArch64 has several condition code registers (`fcc*`) which should be marked as clobbered by default unless `preserves_flags` is set. This should also be documented in the reference.

Finally, a minor detail: the code for `supported_types` in compiler/rustc_target/src/asm/loongarch.rs could be simplified since `arch == InlineAsmArch::LoongArch64` is always true.

**Comment by Amanieu** [comment]
Also, LLVM contains this code to select the base pointer register which is reserved:
```c++
// FIXME: other register?
MCRegister getBPReg() { return LoongArch::R31; }
```

The FIXME seems to indicate that this register might change in the future. To stabilize asm support we need to ensure that the LLVM base pointer register choice is fixed.

**Comment by heiher** [comment]
@Amanieu Thank you for pointing out these issues. I will discuss them with the developers of LLVM and resolve them as soon as possible.

**Comment by heiher** [comment]
@Amanieu I had a conversation with the maintainer of the LoongArch architecture in LLVM on internal OA regarding the reserve of `S8($r31)` as a base pointer (BP), and I received a positive response. It was confirmed that the use of S8 will remain stable and unchanged. Additionally, the `FIXME` comments will be removed.

**Comment by Amanieu** [comment]
LGTM once #111332 is merged. For stabilization you'll also want to open a PR in https://github.com/rust-lang/reference to add the register definitions that were removed from the unstable book.

cc @rust-lang/lang for FCP

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by pnkfelix** [comment]
@rfcbot reviewed

**Comment by heiher** [comment]
@joshtriplett I would like to request the inclusion of this PR in the next `beta` branch. Would you reduce the FCP duration to 6 days, please?
